File: Day11/Part1/labyrinth_3d.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>
#include <tuple>
using namespace std;

// 3D BFS to find minimum cost from (0,0,0) to (n-1,m-1,h-1)
int bfs3DBasic(int n, int m, int h, const vector<vector<vector<int>>>& grid) {
    if (grid[0][0][0] == 1 || grid[h-1][n-1][m-1] == 1) return -1;
    
    // BFS queue: (z, x, y, cost)
    queue<tuple<int,int,int,int>> q;
    q.push({0, 0, 0, 0});
    set<tuple<int,int,int>> visited;
    visited.insert({0, 0, 0});
    
    // 6 directions
    int dz[] = {1, -1, 0, 0, 0, 0};
    int dx[] = {0, 0, 1, -1, 0, 0};
    int dy[] = {0, 0, 0, 0, 1, -1};
    
    while (!q.empty()) {
        auto [z, x, y, cost] = q.front();
        q.pop();
        
        // Check destination
        if (z == h-1 && x == n-1 && y == m-1) return cost;
        
        // Explore neighbors
        for (int i = 0; i < 6; ++i) {
            int nz = z + dz[i], nx = x + dx[i], ny = y + dy[i];
            
            if (nz >= 0 && nz < h && nx >= 0 && nx < n && ny >= 0 && ny < m) {
                if (grid[nz][nx][ny] != 1 && visited.find({nz, nx, ny}) == visited.end()) {
                    visited.insert({nz, nx, ny});
                    q.push({nz, nx, ny, cost + 1});
                }
            }
        }
    }
    
    return -1;
}

int main() {
    // Test 1: simple 2x2x2 cube
    vector<vector<vector<int>>> grid1 = {
        {{0, 0}, {0, 0}},
        {{0, 0}, {0, 0}}
    };
    int res1 = bfs3DBasic(2, 2, 2, grid1);
    if (res1 != 3) return 1;
    
    // Test 2: with wall
    vector<vector<vector<int>>> grid2 = {
        {{0, 0}, {1, 0}},
        {{0, 0}, {0, 0}}
    };
    int res2 = bfs3DBasic(2, 2, 2, grid2);
    if (res2 <= 0) return 1;
    
    // Test 3: impossible
    vector<vector<vector<int>>> grid3 = {
        {{0, 1}, {1, 0}},
        {{1, 0}, {0, 1}}
    };
    int res3 = bfs3DBasic(2, 2, 2, grid3);
    if (res3 != -1) return 1;
    
    cout << "Day11 Part1 labyrinth_3d.cpp: all tests passed" << endl;
    return 0;
}

File: learnLeetCode/Day11/Part1/labyrinth_3d.py
"""
Day 11 — Part 1: Dimensional Labyrinth (3D BFS Basic)

Given a 3D grid (n x m x h), find minimum cost from (0,0,0) to (n-1,m-1,h-1).
- 0 = empty
- 1 = wall
- 2 = portal (ignored in Part 1)

Movement: 6 directions (+/- x, y, z), each step costs 1.

Approach: Standard BFS in 3D space.
Time: O(n * m * h), Space: O(n * m * h)
"""
from typing import List
from collections import deque


def bfs_3d_basic(n: int, m: int, h: int, grid: List[List[List[int]]]) -> int:
    """
    Find minimum steps from (0,0,0) to (n-1,m-1,h-1) in 3D grid.
    Returns -1 if unreachable.
    """
    if grid[0][0][0] == 1 or grid[h-1][n-1][m-1] == 1:
        return -1
    
    # BFS: (z, x, y, cost)
    q = deque([(0, 0, 0, 0)])
    visited = set()
    visited.add((0, 0, 0))
    
    # 6 directions in 3D
    directions = [(1,0,0), (-1,0,0), (0,1,0), (0,-1,0), (0,0,1), (0,0,-1)]
    
    while q:
        z, x, y, cost = q.popleft()
        
        # Check if reached destination
        if z == h-1 and x == n-1 and y == m-1:
            return cost
        
        # Explore neighbors
        for dz, dx, dy in directions:
            nz, nx, ny = z + dz, x + dx, y + dy
            
            # Bounds check
            if 0 <= nz < h and 0 <= nx < n and 0 <= ny < m:
                if (nz, nx, ny) not in visited and grid[nz][nx][ny] != 1:
                    visited.add((nz, nx, ny))
                    q.append((nz, nx, ny, cost + 1))
    
    return -1


if __name__ == "__main__":
    # Simple test: 2x2x2 cube, straight path
    grid1 = [
        [[0, 0], [0, 0]],
        [[0, 0], [0, 0]]
    ]
    # Path: (0,0,0) -> (0,1,0) -> (0,1,1) -> (1,1,1) = 3 steps
    result1 = bfs_3d_basic(2, 2, 2, grid1)
    assert result1 == 3, f"Expected 3, got {result1}"
    
    # Test with wall
    grid2 = [
        [[0, 0], [1, 0]],
        [[0, 0], [0, 0]]
    ]
    result2 = bfs_3d_basic(2, 2, 2, grid2)
    assert result2 > 0, f"Should find a path, got {result2}"
    
    # Test impossible
    grid3 = [
        [[0, 1], [1, 0]],
        [[1, 0], [0, 1]]
    ]
    result3 = bfs_3d_basic(2, 2, 2, grid3)
    assert result3 == -1, f"Should be unreachable, got {result3}"
    
    print("Day11 Part1 labyrinth_3d.py: all tests passed")
